Introduce type definitions + use built-in permutations function + remove some unnecessary statements

EZtouch opened the pull request:
Text borrowed from this answer 🙂: https://stackoverflow.com/questions/32557920/what-are-type-hints-in-python-3-5
# Why Type Hints?

1. **Helps Type Checkers:** By hinting at what type you want the object to be the type checker can easily detect if, for instance, you're passing an object with a type that isn't expected.
2. **Helps with documentation:** A third person viewing your code will know what is expected where, ergo, how to use it without getting them `TypeErrors`.
3. **Helps IDEs develop more accurate and robust tools:** Development Environments will be better suited at suggesting appropriate methods when know what type your object is. You have probably experienced this with some IDE at some point, hitting the `.` and having methods/attributes pop up which aren't defined for an object.
# Why use Static Type Checkers?

**Find bugs sooner:** This is self evident, I believe.
**The larger your project the more you need it:** Again, makes sense. Static languages offer a robustness and control that dynamic languages lack. The bigger and more complex your application becomes the more control and predictability (from a behavioral aspect) you require.
**Large teams are already running static analysis:** I'm guessing this verifies the first two points.

EZtouch committed:
Introduce type definitions + remove some unnecessary statements

EZtouch committed:
Use the faster, built-in permutations function

Because why not 😉

NightShadeI commented:
I agree with the type checker which i will merge later, but I disagree with itertools (it is once of the first things I used when I started working on this project) what it does is generate every permutation and store it as a massive array which you can then iterate over which is definitely more efficient, however as soon as the user exceeds ~11 abilities it starts taking up a massive amount of memory (which can freeze a users computer) I developed a mathematical method to generate a new bar after each one as the solution. :) Let me know if you have any alternatives - however.

EZtouch committed:
Use the generated iterator, no need to recreate the list

Saves x2 the amount of memory usage
By doing this, python will keep in memory only the currently used permutation, not all the permutations.

without itertools 4mb
with itertools 4.2mb
with this commit 4.2mb, literally no impact on memory 😎

EZtouch commented:
@NightShadeI Good catch, I resolved the issue in 7de6b97c997a9ef128729747ce473a060e183939
Now it will keep in memory only the currently used permutation, not all the permutations. 😉👍



You might be interested in learning about the difference between containers, iterables, iterators, and generators.
Borrowed the following from: http://nvie.com/posts/iterators-vs-generators/
**Containers:** Data structures that live in memory, and typically hold all their values in memory, too. Includes list, set, dict, tuple, and str.
**Iterables:** Objects that can return an iterator (with the purpose of returning all of its elements). List and Dict are examples of iterables, they can return the data their elements.
**Iterators:** Objects that produce the next value when next() is called on it. _It is not a data structure that is stored in memory, it performs what is known as lazy loading. From the outside, the iterator is like a lazy factory that is idle until you ask it for a value, which is when it starts to buzz and produce a single value, after which it turns idle again_
**Generators:** Generators are iterators written in a format without writing classes that implement iter() and next()

EZtouch commented:
@NightShadeI Since any irrelevant commit I push to my fork will update this pull request, I will wait until this is merged. After this is merged, I will be opening another pull request where the ability data currently stored in multiple lists will be moved to an sqlite database file, which is cross-platform and doesn't require any installation 😄. We can easily use it to read and store all the ability data in addition to the cached revolution bars. The sqlite3 imported module consumes only an additional 300kb of memory, and is built-in Python3. Reading and populating every list is as fast as the current implementation, if not faster.
I'm working on this project bit by bit to make my changes easier to be understood by everyone. I prefer multiple small changes over one large ambiguous one. 🙂

NightShadeI commented:
Thanks for this looks good 👍 Would be interested to see the sqlite database too!
